R/proj.R

```diff
@@ -0,0 +1,249 @@
+    }
+
+    if (isTRUE(pfdr)) {
+        rlang::check_installed("matrixStats", "for pFDR control.")
+        pd = pd - pa
+        m_pv = matrixStats::rowRanks(abs(pd), ties.method="max") / ncol(pd)
+        pv_draw = matrixStats::rowMeans2(abs(pd) >= abs(contr)) + 1 / ncol(pd)
+
+        attr(contr, "q") = est_pfdr(pv_draw, m_pv) |>
+            qvalues(pv_draw)
+    }
+
+    contr
+}
+
+# Algorithm 1 of Storey & Tibshirani (2001)
+est_pfdr <- function(p, ref, lambda=0.5, n_alpha=15) {
+    alphas = quantile(p, probs=seq(0, 1, length.out=n_alpha), type=1, names=FALSE)
+    m = length(p)
+    pi0 = (m - sum(p <= lambda)) / (m - mean(colSums(ref <= lambda)))
+    pi0 = min(pi0, 1)
+    fdrs = vapply(alphas, function(a) {
+        pi0 *  mean(colSums(ref <= a)) / sum(p <= a)
+    }, 0.0)
+    ok = !is.nan(fdrs)
+
+    list(
+        alpha = alphas,
+        pfdr = dplyr::coalesce(fdrs, 0.0)
+    )
+}
+# Produce q values with spline
+qvalues <- function(ests, p) {
+    idx = vctrs::vec_unique_loc(ests$alpha)
+    splinefun(ests$alpha[idx], rev(cummin(rev(ests$pfdr[idx]))), method="monoH.FC")(p)
+}
+
+
+
+
+#' Plot a Projective Contrast with positive False Discovery Rate (pFDR) Control
+#'
+#' Plot a projective contrast on a map with areas selected by the pFDR control
+#' procedure hatched.
+#'
+#' @param map A [redist_map] object
+#' @param contr The output of [proj_contr()] with `pfdr=TRUE`: A vector containing
+#'   the contrast and an attribute `"q"` containing the q-values.
+#' @param level The positive false discovery rate level to control.
+#' @param density The density of the hatching (roughly what portion is shaded).
+#' @param spacing The spacing of the hatches.
+#'
+#' @returns A ggplot.
+#'
+#' @examples
+#' # example code
+#' data(iowa)
+#' map <- redist_map(iowa, existing_plan = cd_2010, pop_tol = 0.01)
+#' plans <- redist_smc(map, 50, silent = TRUE)
+#' plans$dem <- group_frac(map, dem_08, tot_08, plans)
+#'
+#' pc = proj_contr(plans, dem, pfdr=TRUE)
+#' redist.plot.contr_fdr(map, pc)
+#'
+#'
+#' @concept plot
+#' @export
+redist.plot.contr_fdr <- function(map, contr, level=0.05, density=0.2, spacing=0.015) {
+    if (is.null(attr(contr, "q"))) {
+        cli_abort("Must provide {.arg pfdr=TRUE} to {.fn proj_contr} to use {.fn redist.plot.contr_fdr}.")
+    }
+
+    p = plot(map, contr)
+    q = attr(contr, "q")
+    if (!is.numeric(level) || level < 0 || level > 1) {
+        cli_abort("{.arg level} must be a number between 0 and 1.")
+    }
+    if (!any(q <= level)) return(p)
+    rlang::check_installed("ggpattern", "for pFDR-controlling projective contrast plots.")
+
+    geom_sel = map |>
+        dplyr::filter(q <= level) |>
+        dplyr::summarize(is_coverage=TRUE) |>
+        suppressWarnings()
+
+    p +
+        ggplot2::geom_sf(data=geom_sel, inherit.aes=FALSE,
+                         fill=NA, linewidth=0.15, color="black") +
+        ggpattern::geom_sf_pattern(data=geom_sel, inherit.aes=FALSE,
```

Comment:
I like the ggpattern use, but is it possible to instead do a `rlang::is_installed()` if else, rather than requiring it? Since we only use ggpattern once and it's a much less common package than some of the other things we suggest

Reply:
(a) you have to do `rlang::is_installed()` or `check_installed()` no matter what if you're using a `Suggest`-ed package, as is already done here. And you have to `Suggest` any package you use in this way.  We `Suggest` plenty of one-off packages: `RSpectra`, `Rmpi`, `matrixStats`, etc.

(b) Looked into replacing with new `ggplot2` but sadly pattern support doesn't extend to `geom_sf()` as we'd need